File: web/src/generators/ParametricGridGenerator.js
import { Generator } from '../core/Generator';
import { ParameterDefinition } from '../core/ParameterDefinition';
import { SvgCanvas } from '../core/SvgCanvas';
import { SeededRandom } from '../utils/SeededRandom';

export class ParametricGridGenerator extends Generator {
    getId() {
        return "parametric_grid";
    }

    getDisplayName() {
        return "Parametric Grid";
    }

    getParameterDefinitions() {
        return [
            ParameterDefinition.selection("Preset", "Custom", ["Custom", "Subtle Disruption", "High Chaos", "Dense Matrix"], "Select a predefined style"),
            ParameterDefinition.integer("macroSize", 5, 2, 20, "Macro Grid Size (blocks)"),
            ParameterDefinition.integer("microSize", 10, 2, 30, "Micro Grid Size (squares per block)"),
            ParameterDefinition.doubleVal("maxRotation", 45.0, 0.0, 180.0, "Max Chaos Rotation (degrees)"),
            ParameterDefinition.doubleVal("minScale", 0.2, 0.0, 1.0, "Min Scale at bottom"),
            ParameterDefinition.integer("seed", 1234, 0, 100000, "Random Seed")
        ];
    }

    onParameterChanged(paramName, newValue, currentValues) {
        if (paramName === "Preset" && typeof newValue === "string") {
            const preset = newValue;
            switch (preset) {
                case "Subtle Disruption":
                    currentValues["macroSize"] = 4;
                    currentValues["microSize"] = 8;
                    currentValues["maxRotation"] = 15.0;
                    currentValues["minScale"] = 0.5;
                    currentValues["seed"] = 10;
                    return true;
                case "High Chaos":
                    currentValues["macroSize"] = 6;
                    currentValues["microSize"] = 12;
                    currentValues["maxRotation"] = 90.0;
                    currentValues["minScale"] = 0.1;
                    currentValues["seed"] = 20;
                    return true;
                case "Dense Matrix":
                    currentValues["macroSize"] = 10;
                    currentValues["microSize"] = 20;
                    currentValues["maxRotation"] = 30.0;
                    currentValues["minScale"] = 0.3;
                    currentValues["seed"] = 30;
                    return true;
                case "Custom":
                default:
                    return false;
            }
        }
        if (paramName !== "Preset" && currentValues["Preset"] !== "Custom") {
            currentValues["Preset"] = "Custom";
            return true;
        }
        return false;
    }

    generate(params) {
        const macroSize = params["macroSize"] || 5;
        const microSize = params["microSize"] || 10;
        const maxRotation = params["maxRotation"] !== undefined ? params["maxRotation"] : 45.0;
        const minScale = params["minScale"] !== undefined ? params["minScale"] : 0.2;
        const seed = params["seed"] || 1234;

        const width = params["width"] || 800;
        const height = params["height"] || 800;

        const canvas = new SvgCanvas(width, height, 1);
        const rand = new SeededRandom(seed);

        const macroCellW = width / macroSize;
        const macroCellH = height / macroSize;

        for (let r = 0; r < macroSize; r++) {
            for (let c = 0; c < macroSize; c++) {
                let macroXFactor = c / (macroSize - 1);
                let macroYFactor = r / (macroSize - 1);
                if (macroSize === 1) {
                    macroXFactor = 0;
                    macroYFactor = 0;
                }

                const chaosRange = macroXFactor * maxRotation;
                const scaleBase = 1.0 - (macroYFactor * (1.0 - minScale));

                this.drawMicroGrid(canvas, r, c, macroCellW, macroCellH, microSize, chaosRange, scaleBase, rand);
            }
        }

        return canvas.toSvg();
    }

    drawMicroGrid(canvas, macroRow, macroCol, w, h, microSize, chaosRange, scale, rand) {
        const startX = macroCol * w;
        const startY = macroRow * h;

        const cellW = w / microSize;
        const cellH = h / microSize;
        const padding = cellW * 0.1;
        const sqSize = (cellW - padding * 2) * scale;

        for (let r = 0; r < microSize; r++) {
            for (let c = 0; c < microSize; c++) {
                const cx = startX + c * cellW + cellW / 2.0;
                const cy = startY + r * cellH + cellH / 2.0;

                const rotDeg = (rand.nextDouble() * 2 - 1) * chaosRange;
                const rotRad = rotDeg * Math.PI / 180;

                this.drawRotatedSquare(canvas, cx, cy, sqSize, rotRad);
            }
        }
    }

    drawRotatedSquare(canvas, cx, cy, size, angleRad) {
        const half = size / 2.0;
        const x = [-half, half, half, -half];
        const y = [-half, -half, half, half];

        const cosA = Math.cos(angleRad);
        const sinA = Math.sin(angleRad);

        const rx = [];
        const ry = [];

        for (let i = 0; i < 4; i++) {
            rx[i] = cx + (x[i] * cosA - y[i] * sinA);
            ry[i] = cy + (x[i] * sinA + y[i] * cosA);
        }

        for (let i = 0; i < 4; i++) {
            const next = (i + 1) % 4;
            canvas.addLine(0, rx[i], ry[i], rx[next], ry[next]);
        }
    }
}

File: src/main/java/org/trostheide/generativeArt/ParametricGridGenerator.java
package org.trostheide.generativeArt;

import org.trostheide.generativeArt.core.*;

import java.util.List;
import java.util.Map;
import java.util.Locale;
import java.util.Random;

public class ParametricGridGenerator implements ArtGenerator {

    @Override
    public String getId() {
        return "parametric_grid";
    }

    @Override
    public String getDisplayName() {
        return "Parametric Grid";
    }

    public List<ParameterDefinition> getParameterDefinitions() {
        return List.of(
                ParameterDefinition.selection("Preset", "Custom", 
                    java.util.Arrays.asList("Custom", "Subtle Disruption", "High Chaos", "Dense Matrix"), 
                    "Select a predefined style"),
                ParameterDefinition.integer("macroSize", 5, 2, 20, "Macro Grid Size (blocks)"),
                ParameterDefinition.integer("microSize", 10, 2, 30, "Micro Grid Size (squares per block)"),
                ParameterDefinition.doubleVal("maxRotation", 45.0, 0.0, 180.0, "Max Chaos Rotation (degrees)"),
                ParameterDefinition.doubleVal("minScale", 0.2, 0.0, 1.0, "Min Scale at bottom"),
                ParameterDefinition.integer("seed", 1234, 0, 100000, "Random Seed"));
    }

    @Override
    public boolean onParameterChanged(String paramName, Object newValue, Map<String, Object> currentValues) {
        if ("Preset".equals(paramName) && newValue instanceof String) {
            String preset = (String) newValue;
            switch (preset) {
                case "Subtle Disruption":
                    currentValues.put("macroSize", 4);
                    currentValues.put("microSize", 8);
                    currentValues.put("maxRotation", 15.0);
                    currentValues.put("minScale", 0.5);
                    currentValues.put("seed", 10);
                    return true;
                case "High Chaos":
                    currentValues.put("macroSize", 6);
                    currentValues.put("microSize", 12);
                    currentValues.put("maxRotation", 90.0);
                    currentValues.put("minScale", 0.1);
                    currentValues.put("seed", 20);
                    return true;
                case "Dense Matrix":
                    currentValues.put("macroSize", 10);
                    currentValues.put("microSize", 20);
                    currentValues.put("maxRotation", 30.0);
                    currentValues.put("minScale", 0.3);
                    currentValues.put("seed", 30);
                    return true;
                case "Custom":
                default:
                    return false;
            }
        }
        if (!"Preset".equals(paramName) && !"Custom".equals(currentValues.get("Preset"))) {
            currentValues.put("Preset", "Custom");
            return true;
        }
        return false;
    }

    @Override
    public String generate(Map<String, Object> params) {
        int macroSize = (int) params.getOrDefault("macroSize", 5);
        int microSize = (int) params.getOrDefault("microSize", 10);
        double maxRotation = (double) params.getOrDefault("maxRotation", 45.0);
        double minScale = (double) params.getOrDefault("minScale", 0.2);
        int seed = (int) params.getOrDefault("seed", 1234);

        double width = 800;
        double height = 800; // Square canvas fits this best

        // Single layer for now, usually black on white
        SvgCanvas canvas = new SvgCanvas(width, height, 1);
        canvas.setStrokeWidth(1.5); // Slightly thicker lines look nice

        Random rand = new Random(seed);

        double macroCellW = width / macroSize;
        double macroCellH = height / macroSize;

        for (int r = 0; r < macroSize; r++) {
            for (int c = 0; c < macroSize; c++) {
                // Macro Position Normalized (0.0 to 1.0)
                // Center of randomness vs Edge?
                // Request: Left to Right -> Chaos increases.
                // Request: Top to Bottom -> Size decreases.

                double macroXFactor = (double) c / (macroSize - 1); // 0 at left, 1 at right
                double macroYFactor = (double) r / (macroSize - 1); // 0 at top, 1 at bottom
                if (macroSize == 1) {
                    macroXFactor = 0;
                    macroYFactor = 0;
                }

                // Local parameters for this block
                double chaosRange = macroXFactor * maxRotation; // degrees
                double scaleBase = 1.0 - (macroYFactor * (1.0 - minScale)); // 1.0 -> minScale

                drawMicroGrid(canvas, r, c, macroCellW, macroCellH, microSize, chaosRange, scaleBase, rand);
            }
        }

        return canvas.toSvg();
    }

    private void drawMicroGrid(SvgCanvas canvas, int macroRow, int macroCol,
            double w, double h, int microSize,
            double chaosRange, double scale, Random rand) {

        double startX = macroCol * w;
        double startY = macroRow * h;

        double cellW = w / microSize;
        double cellH = h / microSize;

        // Gap for aesthetics
        double padding = cellW * 0.1;
        double sqSize = (cellW - padding * 2) * scale;

        for (int r = 0; r < microSize; r++) {
            for (int c = 0; c < microSize; c++) {
                double cx = startX + c * cellW + cellW / 2.0;
                double cy = startY + r * cellH + cellH / 2.0;

                // Rotation
                // Chaos: -range to +range
                double rotDeg = (rand.nextDouble() * 2 - 1) * chaosRange;
                double rotRad = Math.toRadians(rotDeg);

                drawRotatedSquare(canvas, cx, cy, sqSize, rotRad);
            }
        }
    }

    private void drawRotatedSquare(SvgCanvas canvas, double cx, double cy, double size, double angleRad) {
        double half = size / 2.0;

        // Corners relative to center (unrotated)
        double[] x = { -half, half, half, -half };
        double[] y = { -half, -half, half, half };

        double cosA = Math.cos(angleRad);
        double sinA = Math.sin(angleRad);

        double[] rx = new double[4];
        double[] ry = new double[4];

        for (int i = 0; i < 4; i++) {
            rx[i] = cx + (x[i] * cosA - y[i] * sinA);
            ry[i] = cy + (x[i] * sinA + y[i] * cosA);
        }

        // Draw 4 lines
        for (int i = 0; i < 4; i++) {
            int next = (i + 1) % 4;
            canvas.addLine(0, rx[i], ry[i], rx[next], ry[next]);
        }
    }
}
